addons/frag/functions/fnc_doSpall.sqf

```diff
@@ -1,139 +1,146 @@
+private _deltaStep = _lastVelocityUnit vectorMultiply 0.05;
 
-    private _spallFragVect = [_vel, _dir, _elev] call CBA_fnc_polar2vect;
-    private _fragment = (selectRandomWeighted WEIGHTED_SIZE) createVehicleLocal [0,0,10000];
-    _fragment setPosASL _spallPos;
-    _fragment setVelocity _spallFragVect;
+if (terrainIntersectASL [_lastPosASL vectorAdd _deltaStep, _lastPosASL]) exitWith {
+    TRACE_2("terrainIntersect",_lastPosASL,_deltaStep);
+};
 
-    #ifdef DRAW_FRAG_INFO
-        [ACE_player, _fragment, [1, 0.5, 0, 1]] call FUNC(dev_addTrack);
-    #endif
+#ifdef DEBUG_MODE_DRAW
+if GVAR(dbgSphere) then {
+    [_lastPosASL vectorAdd _lastVelocityUnit, "orange"] call FUNC(dev_sphereDraw);
+    [_lastPosASL, "yellow"] call FUNC(dev_sphereDraw);
 };
+#endif
 
-_spread = 5 + (random 5);
-_spallCount = 3 + (random 5);
-for "_i" from 1 to _spallCount do {
-    private _elev = ((_spallPolar select 2) - _spread) + (random (_spread * 2));
-    private _dir = ((_spallPolar select 1) - _spread) + (random (_spread * 2));
-    if (abs _elev > 90) then {
-        ADD(_dir,180);
+/*
+ * Improve performance of finding otherside of object on shallow angle
+ * impacts. 120 degrees due to 90 degree offset with _lastVelocityUnit into object.
+ */
+private _spallPosASL = _lastPosASL vectorAdd _deltaStep;
+if (120 > acos ( _lastVelocityUnit vectorDotProduct _surfaceNorm)) then {
+    _spallPosASL = _spallPosASL vectorAdd (_deltaStep vectorMultiply 5);
+};
+private _insideObject = true;
+for "_i" from 2 to 21 do
+{
+    private _nextPos = _spallPosASL vectorAdd _deltaStep;
+    if (!lineIntersects [_spallPosASL, _nextPos]) then {
+        _spallPosASL = _nextPos vectorAdd (_deltaStep vectorMultiply 2);
+        _insideObject = false;
+        break
     };
-    _dir = _dir % 360;
-    private _vel = (_spallPolar select 0) * 0.55 * _vm;
-    _vel = (_vel - (_vel * 0.25)) + (random (_vel * 0.5));
+    _spallPosASL = _nextPos;
+};
 
-    private _spallFragVect = [_vel, _dir, _elev] call CBA_fnc_polar2vect;
-    private _fragment = (selectRandomWeighted WEIGHTED_SIZE) createVehicleLocal [0, 0, 10000];
-    _fragment setPosASL _spallPos;
-    _fragment setVelocity _spallFragVect;
+if (_insideObject) exitWith {
+    TRACE_3("insideObj",_lastPosASL,_spallPosASL,alive _projectile);
+};
+// Passed all exitWiths
+GVAR(nextSpallAllowTime) = CBA_missionTime + ACE_FRAG_SPALL_HOLDOFF;
 
-    #ifdef DRAW_FRAG_INFO
-        [ACE_player, _fragment, [1, 0, 0, 1]] call FUNC(dev_addTrack);
-    #endif
+#ifdef DEBUG_MODE_DRAW
+if GVAR(dbgSphere) then {
+    [_spallPosASL, "green"] call FUNC(dev_sphereDraw);
+};
+#endif
+
+private _spawnSize = switch (true) do
+{
+    case (_spallPower < 3): { "_spall_tiny" };
+    case (_spallPower < 5): { "_spall_small" };
+    case (_spallPower < 8): { "_spall_medium" };
+    case (_spallPower < 12): { "_spall_large" };
+    default { "_spall_huge" };
 };
+
+private _spallSpawner = createVehicle [
+    "ace_frag_" + _material + _spawnSize,
```

Comment:
1. It's faster to use string addition in this case, unless I'm missing something?
2. Is there a way to do QUOTE(ADDON)?

Reply:
1. The idea is to use the macros, so in case one day we want to change the component, it's much easier to do.
2. QUOTE(ADDON) exists, you can do `QUOTE(ADDON) + "_"  + _material + _spawnSize`. I imagine it's still faster than format, so go with that.

Reply:
Going with this: `QUOTE(GLUE(ADDON,_)) + _material + _spawnSize`

Result:
0.00121526 ms

Cycles:
10000/10000

Code:
private _material = "ground";
private _spawnSize = "_spall_large";
"ace_frag_" + _material + _spawnSize

Result:
0.00136958 ms

Cycles:
10000/10000

Code:
private _material = "ground";
private _spawnSize = "_spall_large";
format ["ace_frag_%1%2",_material,_spawnSize]

Reply:
done by https://github.com/acemod/ACE3/pull/9728/commits/66c566b772c9ff7f7a66a77e86f909b0fcab9f67